src/large_pages/node_large_page.cc

```diff
@@ -144,43 +132,120 @@ inline uintptr_t hugepage_align_down(uintptr_t addr) {
+  FORCE_INLINE ~MemoryMapPointer() {
+    if (mem_ == nullptr) return;
+    if (mem_ == MAP_FAILED) return;
+    if (munmap(mem_, size_) == 0) return;
+    PrintSystemError(errno);
+  }
+
+ protected:
+  FORCE_INLINE void Reset(void* data, size_t size) {
+    mem_ = data;
+    size_ = size;
+  }
+
+ private:
+  size_t size_ = 0;
+  void* mem_ = nullptr;
+};
+
+class MappedFilePointer: public MemoryMapPointer {
+ public:
+  CONSTRUCT_ONLY(MappedFilePointer);
+  FORCE_INLINE explicit MappedFilePointer() {}
+  FORCE_INLINE void Reset(void* start,
+                          const char* fname,
+                          int prot,
+                          int flags,
+                          size_t offset = 0) {
+    struct stat file_info;
+    if (stat(fname, &file_info) == -1) goto fail;
+
+    fd_ = open(fname, O_RDONLY);
+    if (fd_ == -1) goto fail;
+
+    MemoryMapPointer::Reset(start, file_info.st_size, prot, flags, fd_, offset);
+    return;
+fail:
+    MemoryMapPointer::Reset(MAP_FAILED, 0);
+  }
+
+  FORCE_INLINE ~MappedFilePointer() {
+    if (fd_ == -1) return;
+    if (close(fd_) == 0) return;
+    PrintSystemError(errno);
+  }
+
+ private:
+  int fd_ = -1;
+};
+
 struct dl_iterate_params {
-  uintptr_t start;
-  uintptr_t end;
-  uintptr_t reference_sym;
+  uintptr_t base_address;
   std::string exename;
 };
 
-int FindMapping(struct dl_phdr_info* info, size_t, void* data) {
+int FindBaseAddress(struct dl_phdr_info* info, size_t, void* data) {
   auto dl_params = static_cast<dl_iterate_params*>(data);
   if (dl_params->exename == std::string(info->dlpi_name)) {
-    for (int idx = 0; idx < info->dlpi_phnum; idx++) {
-      const ElfW(Phdr)* phdr = &info->dlpi_phdr[idx];
-      if (phdr->p_type == PT_LOAD && (phdr->p_flags & PF_X)) {
-        uintptr_t start = info->dlpi_addr + phdr->p_vaddr;
-        uintptr_t end = start + phdr->p_memsz;
-
-        if (dl_params->reference_sym >= start &&
-            dl_params->reference_sym <= end) {
-          dl_params->start = start;
-          dl_params->end = end;
-          return 1;
-        }
-      }
-    }
+    dl_params->base_address = info->dlpi_addr;
+    return 1;
   }
   return 0;
 }
 #endif  // defined(__linux__) || defined(__FreeBSD__)
 
+bool FindTextSection(const std::string& exename, ElfW(Shdr)* text_section) {
+  MappedFilePointer exe;
+  exe.Reset(nullptr, exename.c_str(), PROT_READ, MAP_PRIVATE);
+  if (exe.mem() == MAP_FAILED) return false;
+
+  const char* exe_start = static_cast<char*>(exe.mem());
+  const ElfW(Ehdr)* ehdr = reinterpret_cast<const ElfW(Ehdr)*>(exe_start);
+  const ElfW(Shdr)* shdrs =
+      reinterpret_cast<const ElfW(Shdr)*>(exe_start + ehdr->e_shoff);
+  const char* snames = exe_start + shdrs[ehdr->e_shstrndx].sh_offset;
+
+  for (uint32_t idx = 0; idx < ehdr->e_shnum; idx++) {
+    const ElfW(Shdr)* sh = &shdrs[idx];
+    const char* section_name = static_cast<const char*>(&snames[sh->sh_name]);
+    if (std::string(section_name) == ".text") {
```

Comment:
```suggestion
    if (0 == strcmp(section_name, ".text")) {
```
It's kind of wasteful to create a heap-allocated throwaway string. It's probably going to be stored inline for short strings but you get my point.